File: typescript/exercises.ts

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// First then apply function (Completed with the help of Dr. Toal)
export function firstThenApply<T, U>(
  items: T[],
  predicate: (item: T) => boolean,
  consumer: (item: T) => U
): U | undefined {
  for (const item of items) {
    if (predicate(item)) {
      return consumer(item)
    }
  }
  return undefined
}

// Powers generator function (Completed with the help of Dr. Toal)

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power
  }
}

// Line count function  (Completed with the help of built-in AI tools)
export function meaningfulLineCount(code: string): number {
  let count = 0
  for (const line of code.split("\n")) {
    if (line.trim() !== "" && !line.startsWith("//")) {
      count++
    }
  }
  return count
}

// Shape type and associated functions (Completed with the help of Dr. Toal)
interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
    case "Box":
      return 2 * (shape.width * shape.length + shape.length * shape.depth + shape.depth * shape.width)
  }
}

export function volume(shape: Shape): number { 
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3
    case "Box":
      return shape.width * shape.length * shape.depth
  }
}

// Binary Search Tree (Completed with the help of built-in AI tools)
export interface BinarySearchTree<T> {
  size(): number;
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  inorder(): T[];
  toString(): string;
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0;
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty());
  }

  contains(value: T): boolean {
    return false;
  }

  inorder(): T[] {
    return [];
  }

  toString(): string {
    return "";
  }
}

class Node<T> implements BinarySearchTree<T> {
  value: T;
  left: BinarySearchTree<T>;
  right: BinarySearchTree<T>;

  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    this.value = value;
    this.left = left;
    this.right = right;
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right);
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value));
    } else {
      return this;
    }
  }

  contains(value: T): boolean {
    if (value === this.value) {
      return true;
    } else if (value < this.value) {
      return this.left.contains(value);
    } else {
      return this.right.contains(value);
    }
  }

  inorder(): T[] {
    return [...this.left.inorder(), this.value, ...this.right.inorder()];
  }

  toString(): string {
    const leftString = this.left.toString();
    const rightString = this.right.toString();
    const leftPart = leftString ? leftString : "";
    const rightPart = rightString ? rightString : "";
    return `(${leftPart}${this.value}${rightPart})`;
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* First Then apply function done with Homework Helper 3 *)
let first_then_apply array predicate customer =
  match List.find_opt predicate array with
  | None -> None
  | Some x -> customer x

(* Powers generator done with Homework Helper 3 *)
let powers_generator base =
  let rec generate_from power () =
    Seq.Cons (power, generate_from (power * base))
  in
  generate_from 1

(* Line count function from Dr. Toal in office *)
let meaningful_line_count filename =

  let is_meaningful_line line =
    let trimmed = String.trim line in
    trimmed <> "" && not (String.starts_with ~prefix:"#" trimmed)
  in

  let ic = open_in filename in
  let finally () = close_in ic in

  let rec count_lines count =
    match input_line ic with
    | line ->
      let new_count = if is_meaningful_line line then count + 1 else count in
      count_lines new_count
    | exception End_of_file ->
      count
  in
  Fun.protect ~finally (fun () -> count_lines 0)

(* Shape type and associated functions *)
type shape =
  | Sphere of float
  | Box of float * float * float

let volume s = 
  match s with
  | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
  | Box (l, w, h) -> l *. w *. h

let surface_area s = (* from Dr. Toal in office *)
  match s with
  | Sphere r -> 4.0 *. Float.pi *. (r ** 2.0)
  | Box (w, l, d) -> 2.0 *. ((w *. l) +. (w *. d) +. (l *. d))


(* Binary search tree implementation done with help from built-in AI tools *)
type 'a binary_search_tree =
  | Empty 
  | Node of 'a binary_search_tree * 'a *  'a binary_search_tree

let rec size tree =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec contains value tree =
  match tree with
  | Empty -> false
  | Node (left, v, right) -> 
    if value = v then
      true
    else if value < v then
      contains value left
    else
      contains value right

let rec inorder tree =
  match tree with
  | Empty -> []
  | Node (left, v, right) -> inorder left @ [v] @ inorder right

let insert value tree =
  let rec aux value tree =
    match tree with
    | Empty -> Node (Empty, value, Empty)
    | Node (left, v, right) -> 
      if value = v then
        tree
      else if value < v then
        Node (aux value left, v, right)
      else
        Node (left, v, aux value right)
  in
  aux value tree
